core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java

```diff
@@ -498,31 +506,64 @@ private Query getPossiblyAnalyzedPrefixQuery(String field, String termStr) throw
                 return super.getPrefixQuery(field, termStr);
             }
             tlist = new ArrayList<>();
+            List<String> currentPos = new ArrayList<>();
             CharTermAttribute termAtt = source.addAttribute(CharTermAttribute.class);
+            PositionIncrementAttribute posAtt = source.addAttribute(PositionIncrementAttribute.class);
 
             while (true) {
                 try {
                     if (!source.incrementToken()) break;
                 } catch (IOException e) {
                     break;
                 }
-                tlist.add(termAtt.toString());
+                if (currentPos.isEmpty() == false && posAtt.getPositionIncrement() > 0) {
+                    tlist.add(currentPos);
+                    currentPos = new ArrayList<>();
+                }
+                currentPos.add(termAtt.toString());
+            }
+            if (currentPos.isEmpty() == false) {
+                tlist.add(currentPos);
             }
         } finally {
             if (source != null) {
                 IOUtils.closeWhileHandlingException(source);
             }
         }
 
-        if (tlist.size() == 1) {
-            return super.getPrefixQuery(field, tlist.get(0));
+
+        if (tlist.size() == 1 && tlist.get(0).size() == 1) {
+            return super.getPrefixQuery(field, tlist.get(0).get(0));
         } else {
-            // build a boolean query with prefix on each one...
+            // build a boolean query with prefix on the last position only.
             List<BooleanClause> clauses = new ArrayList<>();
-            for (String token : tlist) {
-                clauses.add(new BooleanClause(super.getPrefixQuery(field, token), BooleanClause.Occur.SHOULD));
+            for (int i = 0; i < tlist.size(); i++) {
+                List<String> plist = tlist.get(i);
+                boolean isLastPos = (i == tlist.size()-1) ? true : false;
+                Query posQuery;
+                if (plist.size() == 1) {
+                    if (isLastPos) {
+                        posQuery = getPrefixQuery(field, plist.get(0));
+                    } else {
+                        posQuery = newTermQuery(new Term(field, plist.get(0)));
+                    }
+                } else {
+                    List<BooleanClause> innerClauses = new ArrayList<>();
+                    for (String token : plist) {
+                        if (isLastPos) {
+                            innerClauses.add(new BooleanClause(getPrefixQuery(field, token),
+                                BooleanClause.Occur.SHOULD));
+                        } else {
+                            innerClauses.add(new BooleanClause(newTermQuery(new Term(field, token)),
+                                BooleanClause.Occur.SHOULD));
+                        }
+                    }
+                    posQuery = getBooleanQueryCoordDisabled(innerClauses);
```

Comment:
if all sub queries are term queries with the same boost, it would be nice to build a SynonymQuery instead